src/build.cc

```diff
@@ -789,7 +805,7 @@ bool Builder::Build(string* err) {
   while (plan_.more_to_do()) {
     // See if we can start any more commands.
     if (failures_allowed) {
-      size_t capacity = command_runner_->CanRunMore();
+      size_t capacity = command_runner_->CanRunMore(plan_.JobserverEnabled());
       while (capacity > 0) {
```

Comment:
Please correct me if I am wrong, but it looks like that now that CanRunMode() returns SIZE_MAX unconditionally when jobserver is enabled, this loop will spawn immediately _any_ ready command (unless they are in a pool with a specific depth), regardless of the number of already running jobs, and cpu load.

This seems to assume that all these commands will recognize the MAKEFLAGS environment variables and respect its contract, and not each use a lot of RAM and CPU before actually doing so. I don't think this is generally true.

I fear that for large build graphs, this can lead to hundreds if not thousands of commands being spawned in parallel with no control whatsoever.

Am I understanding things correctly, or is there another mechanism being used here to limit process explosion?

Reply:
I think I found it: FindWork() will return null if the token could not be acquired, so this effectively limits the number of processes that Ninja will ever spawn. And the Acquire() / Release() methods are never blocking (related to my other comment in jobserver.h).

Now I am curious to understand what Ninja does where there are no spawned processes anymore, and no tokens available. Do you know if Ninja would be busy-looping in this case?

Reply:
Thanks, replying here to your email answer, which mentions that the outer loop ends up blocking in SubprocessSet::DoWork(). It looks like, from the implementation, that if there is no running commands, this would either block infintely or just busy-loop calling `perror("ppoll")`, depending on whether ppoll() returns an error where there are no fds and no timeout provided to the syscall.

Neither is really good, but this is probably an edge case that we should document somewhere, and shouldn't stop submitting this PR. And please don't be mistaken by my remarks, this is truly an excellent feature, so thank you for uploading it :-)

Server mode, and client-server passthrough will probably require much more complicated changes, but this hits the sweet spot for client-only support!

Reply:
> Thanks, replying here to your email answer, which mentions that the outer loop ends up blocking in SubprocessSet::DoWork(). It looks like, from the implementation, that if there is no running commands, this would either block infintely or just busy-loop calling `perror("ppoll")`, depending on whether ppoll() returns an error where there are no fds and no timeout provided to the syscall.

It can only go into `WaitForCommand()` if `pending_commands` is non-zero, so we shouldn't end up stuck. On the other hand, if `FindWork()` returns null and `pending_commands` is zero, then ninja fails with `"stuck [this is a bug]"`. But that also doesn't happen, because the first token is always available. Think of it as the main ninja process being handed a token from the caller when started.